llvm/test/tools/sycl-post-link/spec-constants/default-value/device-image.ll

```diff
@@ -0,0 +1,44 @@
+; Test checks the content of simple generated device image.
+
+; RUN: sycl-post-link -split=auto -spec-const=rt -symbols -S -o %t.table %s -generate-device-image-default-spec-consts
+; RUN: cat %t.table | FileCheck %s -check-prefix=CHECK-TABLE -DPATH=%t
+; RUN: cat %t_0.prop | FileCheck %s -check-prefix=CHECK-PROP0
+; RUN: cat %t_1.prop | FileCheck %s -check-prefix=CHECK-PROP1
+; RUN: cat %t_0.ll | FileCheck %s -check-prefix=CHECK-IR0
+; RUN: cat %t_1.ll | FileCheck %s -check-prefix=CHECK-IR1
+
+; CHECK-TABLE: [[PATH]]_0.ll|[[PATH]]_0.prop|[[PATH]]_0.sym
+; CHECK-TABLE: [[PATH]]_1.ll|[[PATH]]_1.prop|[[PATH]]_1.sym
+
+; CHECK-PROP0-NOT: defaultSpecConstants=1|1
+; CHECK-PROP0-NOT: originalImage
+
+; CHECK-PROP1: defaultSpecConstants=1|1
+
+; CHECK-IR0: call i32 @_Z20__spirv_SpecConstantii
+; CHECK-IR0: call %struct.B @_Z29__spirv_SpecConstantCompositeiii_Rstruct.B
+; CHECK-IR0: call %struct.A @_Z29__spirv_SpecConstantCompositeistruct.B_Rstruct.A
+
+; CHECK-IR1-NOT: SpecConstant
+; CHECK-IR1: call void @llvm.memcpy.p4i8.p4i8.i64(i8 addrspace(4)* align 8 %1, i8 addrspace(4)* align 8 addrspacecast (i8 addrspace(1)* bitcast (%"class.sycl::_V1::specialization_id" addrspace(1)* @_ZL1c to i8 addrspace(1)*) to i8 addrspace(4)*), i64 16, i1 false)
```

Comment:
Have you performed experiments to ensure that this `memcpy` can successfully be optimized out by some constant propagation passes? Because if not, then we can't emit it here.

Moreover, I would suggest not to emit it right away, but instead to plain `store` of a constant literal value, which we should be able to obtain from the global variable

Reply:
Moved to `store` instruction instead of `memcpy`.